### PR TITLE
Validate forge command outputs in docs

### LINKED ISSUE #2684: Improve skipping docs snippets
### Which component is your feature related to?

Other (Please describe below)

### Feature Request

Currently, skipped docs snippets are hardcoded in tests (respective `test_docs_snippets` for sncast/forge). This should be improved, potentially e.g. by annotating docs snippet as ignored/skipped in markdown.

### PR DESCRIPTION
**Pull request by franciszekjob**
Relates #2479 
Closes #2684

## Introduced changes

- Automatically run snforge commands which are present in the docs; if the output of respective output is present in the docs too,  it's compared with the actual output from executed command
- Allow ignoring commands (so they will not be executed)

## Checklist

- [x] Linked relevant issue
- [x] Updated relevant documentation
- [x] Added relevant tests
- [x] Performed self-review of the code
- [x] Added changes to `CHANGELOG.md`

### CODE REVIEW

File: scripts/verify_cairo_listings.sh [hunk 1]
```diff
@@ -1,4 +1,4 @@
 #!/bin/bash
 set -e
 
-for d in ./docs/listings/*; do (cd "$d" && scarb test); done
+for d in ./docs/listings/*; do (cd "$d" && scarb check); done
```
Comment (franciszekjob) [inline scripts/verify_cairo_listings.sh:4]: note: Instead of calling `scarb test` we should call `scarb check`, some tests in examples are failing on purpose.

**Comment by franciszekjob** [comment]
Validation of sncast command outputs will be added in the next PR.

**Comment by franciszekjob** [comment]
For some reason, `Test Building Docs` fails on CI ([example here](https://github.com/foundry-rs/starknet-foundry/actions/runs/11949694261/job/33309668448)). It tries to open a scarb plugin file which doesn't exist 🤔 .

File: crates/forge/tests/e2e/docs_snippets_validation.rs [hunk 2]
```diff
@@ -1,43 +1,96 @@
...
+            let package_name = snippet
+                .capture_package_from_output()
+                .expect("Failed to capture package from command output");
+
+            let temp =
+                if is_package_from_docs_listings(&package_name, &listings_to_packages_mapping) {
+                    setup_package_from_docs_listings(&package_name, &listings_to_packages_mapping)
+                } else {
+                    resolve_package_name(&package_name)
+                };
```
Comment (franciszekjob) [inline crates/forge/tests/e2e/docs_snippets_validation.rs:80]: note: Currently we use `hello_workpaces` package (from `crates/forge/tests/data` in one example in docs. In future we can simplify this logic, once we use only packages from docs listings.

File: crates/docs/src/validation.rs [hunk 3]
```diff
@@ -90,26 +142,75 @@ pub fn get_parent_dir(levels_up: usize) -> PathBuf {
...
+                }
+            }
+        }
+    }
+
+    mapping
+}
+
+fn list_packages_in_directory(dir_path: &Path) -> Vec<String> {
+    let crates_path = dir_path.join("");
```
Comment (Draggu) [inline crates/docs/src/validation.rs:200]:
```suggestion
    let crates_path = dir_path.to_owned();
```

Also why you need this at all?

File: crates/docs/src/validation.rs [hunk 4]
```diff
@@ -1,15 +1,46 @@
...
+    collections::HashMap,
     env, fs, io,
     path::{Path, PathBuf},
 };
+use walkdir::WalkDir;
 
 const EXTENSION: Option<&str> = Some("md");
 
+#[derive(Clone, Debug)]
+pub enum SnippetType {
```
Comment (Draggu) [inline crates/docs/src/validation.rs:12]: Should be more like 

```rs
pub struct SnippetType(String);
```

with convenient constructors like `SnippetType::snforge()`

docs should not be based on 2 tools that are using this crate

File: crates/forge/tests/e2e/docs_snippets_validation.rs [hunk 5]
```diff
@@ -1,43 +1,91 @@
...
+    listings_to_packages_mapping: &HashMap<String, Vec<String>>,
+) -> TempDir {
+    if is_package_from_docs_listings(package_name, listings_to_packages_mapping) {
+        setup_package_from_docs_listings(package_name, listings_to_packages_mapping)
+    } else if ["addition", "fibonacci"].contains(&package_name) {
+        setup_hello_workspace()
+    } else {
+        setup_package(package_name)
+    }
 }
```
Comment (franciszekjob) [inline crates/forge/tests/e2e/docs_snippets_validation.rs:91]: note: Currently we use `hello_workpaces` package (from `crates/forge/tests/data`) in one example in docs. In future we can simplify this logic and remove `is_package_from_docs_listings`, once we use only packages from docs listings. Ref: https://github.com/foundry-rs/starknet-foundry/issues/2698

File: crates/docs/src/validation.rs (hunk 3 above)
Comment (franciszekjob) [reply crates/docs/src/validation.rs:200]: It's used to retrieve packages located in docs listing directory.

File: crates/sncast/tests/docs_snippets/validation.rs [hunk 6]
```diff
@@ -63,8 +46,8 @@ fn test_docs_snippets() {
         let exit_code = output.status.code().unwrap_or_default();
         let stderr = String::from_utf8_lossy(&output.stderr);
 
-        assert_valid_snippet(exit_code != 2, snippet, "sncast", &stderr);
+        assert_valid_snippet(exit_code != 2, snippet, &stderr);
```
Comment (cptartur) [inline crates/sncast/tests/docs_snippets/validation.rs:49]: Why not check if it's `0`?

File: scripts/verify_cairo_listings.sh (hunk 1 above)
Comment (cptartur) [reply scripts/verify_cairo_listings.sh:4]: How did it work before?

File: crates/forge/tests/e2e/common/runner.rs [hunk 7]
```diff
@@ -60,6 +66,12 @@ pub(crate) fn setup_package_with_file_patterns(
         value(get_assert_macros_version().unwrap().to_string());
     scarb_toml["target.starknet-contract"]["sierra"] = value(true);
 
+    if package_path.contains("docs/listings") {
+        scarb_toml["dev-dependencies"]["snforge_std"]
+            .as_table_mut()
+            .and_then(|snforge_std| snforge_std.remove("workspace"));
+    }
```
Comment (cptartur) [inline crates/forge/tests/e2e/common/runner.rs:73]: Don't think we should be hardcoding some docs specific behaviors here. This could be a separate function that shares the logic with this one. 

\+ I think you wouldn't have to change the interface of current `setup_package_with_file_patterns` as well.

Comment (cptartur) [reply crates/forge/tests/e2e/common/runner.rs:73]: I see that you created one below, but I'd try to change this so the interface of tests we already had remains unchanged.

File: crates/docs/src/validation.rs [hunk 8]
```diff
@@ -90,26 +149,75 @@ pub fn get_parent_dir(levels_up: usize) -> PathBuf {
...
+            if let Some(listing_name) = listing_path.file_name().and_then(|x| x.to_str()) {
+                if !packages.is_empty() {
+                    mapping.insert(listing_name.to_string(), packages);
+                }
+            }
+        }
+    }
+
+    mapping
+}
```
Comment (cptartur) [inline crates/docs/src/validation.rs:201]: Isn't this logic a bit too complicated just to get a mapping between a directory name and it's related crates?

File: crates/docs/src/validation.rs [hunk 9]
```diff
@@ -1,15 +1,51 @@
...
+    #[must_use]
+    pub fn as_str(&self) -> &str {
+        &self.0
+    }
+
+    #[must_use]
+    pub fn get_re(&self) -> Regex {
+        let escaped_command = regex::escape(self.as_str());
+        let pattern = format!(
+            r"(?ms)^(?:<!--\s*(.*?)\s*-->\n)?```shell\n\$ ({escaped_command} .+?)\n```(?:\s*<details>\n<summary>Output:<\/summary>\n\n```shell\n([\s\S]+?)\n```[\s]*<\/details>)?"
```
Comment (ddoktorski) [inline crates/docs/src/validation.rs:34]: Could you add a short comment explanation of this regex?

File: crates/forge/tests/e2e/common/runner.rs [hunk 10]
```diff
@@ -69,7 +81,29 @@ pub(crate) fn setup_package_with_file_patterns(
...
+) -> Option<String> {
+    packages_mapping
+        .iter()
+        .find(|(_, packages)| packages.contains(&package.to_string()))
+        .map(|(listing_name, _)| listing_name.clone())
+}
+
+pub(crate) fn setup_package_from_docs_listings(
+    package_name: &str,
+    packages_mapping: &HashMap<String, Vec<String>>,
```
[the remark this answers is not in the document]
Comment (franciszekjob) [reply crates/forge/tests/e2e/common/runner.rs]: It has been completely removed due to refactor.

File: crates/docs/src/validation.rs [hunk 11]
```diff
@@ -90,26 +149,75 @@ pub fn get_parent_dir(levels_up: usize) -> PathBuf {
...
+                    mapping.insert(listing_name.to_string(), packages);
+                }
+            }
+        }
+    }
+
+    mapping
+}
+
+fn list_packages_in_directory(dir_path: &Path) -> Vec<String> {
```
[the remark this answers is not in the document]
Comment (franciszekjob) [reply crates/docs/src/validation.rs]: Removed due to refactor.

File: crates/docs/src/validation.rs [hunk 12]
```diff
@@ -90,26 +149,75 @@ pub fn get_parent_dir(levels_up: usize) -> PathBuf {
...
     println!(
         "Skipped validation of {} snippet in the docs in file: {} at line {}",
-        tool_name, snippet.file_path, snippet.line_start
+        snippet.snippet_type.as_str(),
+        snippet.file_path,
+        snippet.line_start,
     );
 }
+
+pub fn create_listings_to_packages_mapping() -> HashMap<String, Vec<String>> {
```
[the remark this answers is not in the document]
Comment (franciszekjob) [reply crates/docs/src/validation.rs]: Removed due to refactor.

File: crates/docs/src/validation.rs [hunk 13]
```diff
@@ -28,33 +64,56 @@ impl Snippet {
...
-    let snippets = re
+    let snippets = snippet_type
+        .get_re()
         .captures_iter(&content)
         .filter_map(|caps| {
-            let command_match = caps.get(1)?;
+            let command_match = caps.get(2)?;
             let match_start = caps.get(0)?.start();
+            let output = caps.get(3).map(|m| m.as_str().to_string());
+            let ignored = caps.get(1).map_or(false, |m| m.as_str().contains("ignore"));
```
[the remark this answers is not in the document]
Comment (franciszekjob) [reply crates/docs/src/validation.rs]: Reordered.

File: crates/docs/src/validation.rs (hunk 8 above)
Comment (franciszekjob) [reply crates/docs/src/validation.rs]: Removed dur to refactor.

File: crates/forge/tests/e2e/docs_snippets_validation.rs (hunk 5 above)
Comment (franciszekjob) [reply crates/forge/tests/e2e/docs_snippets_validation.rs]: No longer relevant.

File: crates/forge/tests/e2e/common/runner.rs [hunk 14]
```diff
@@ -69,7 +81,29 @@ pub(crate) fn setup_package_with_file_patterns(
...
+    package: &str,
+    packages_mapping: &HashMap<String, Vec<String>>,
+) -> Option<String> {
+    packages_mapping
+        .iter()
+        .find(|(_, packages)| packages.contains(&package.to_string()))
+        .map(|(listing_name, _)| listing_name.clone())
+}
+
+pub(crate) fn setup_package_from_docs_listings(
```
[the remark this answers is not in the document]
Comment (franciszekjob) [reply crates/forge/tests/e2e/common/runner.rs]: No longer relevant.

[the remark this answers is not in the document]
Comment (franciszekjob) [reply crates/forge/tests/e2e/common/runner.rs]: This function has been removed, now we only use `setup_package`.

File: crates/docs/src/validation.rs (hunk 9 above)
Comment (franciszekjob) [reply crates/docs/src/validation.rs:45]: Sure, added.

File: crates/sncast/tests/docs_snippets/validation.rs (hunk 6 above)
Comment (franciszekjob) [reply crates/sncast/tests/docs_snippets/validation.rs:47]: As I mentioned in [this comment](https://github.com/foundry-rs/starknet-foundry/pull/2687#issuecomment-2490793232), sncast commands output validation will be added in the next PR. Currently, we simply check if clap didn't raise any errors, i.e. error code is different than 2.

File: scripts/verify_cairo_listings.sh (hunk 1 above)
Comment (franciszekjob) [reply scripts/verify_cairo_listings.sh:4]: Wdym exactly 😅 ? Some packages have tests which fail on purpose, so the old `scarb test` would result in failure of this script, hope it clarifies.

**Comment by franciszekjob** [comment]
I made a refactor which reorganized all packages used in the docs listings. Now all packages are located under `docs/listings`.

File: docs/listings/panicking_test/src/lib.cairo [hunk 15]
```diff
@@ -11,7 +11,6 @@ mod tests {
 
     #[test]
     //ANCHOR_END:first_half
-    #[should_panic(expected: 'aaa')]
```
Comment (kkawula) [inline docs/listings/panicking_test/src/lib.cairo:14]: why this was deleted?

File: docs/listings/error_handling/Scarb.toml [hunk 16]
```diff
@@ -6,10 +6,10 @@ edition = "2024_07"
 # See more keys and their definitions at https://docs.swmansion.com/scarb/docs/reference/manifest.html
 
 [dependencies]
-snforge_std.workspace = true
-sncast_std.workspace = true
-starknet.workspace = true
-assert_macros.workspace = true
```
Comment (kkawula) [inline docs/listings/error_handling/Scarb.toml:12]: What's the reason for this change? Applies also to other similar changes.

File: docs/listings/panicking_test/src/lib.cairo (hunk 15 above)
Comment (kkawula) [reply docs/listings/panicking_test/src/lib.cairo:14]: shouldn't it be
#[should_panic(expected: 'panic message')]

File: docs/src/testing/running-tests.md [hunk 17]
```diff
@@ -23,8 +24,9 @@ Tests: 3 passed, 0 failed, 0 skipped, 0 ignored, 0 filtered out
 ## Filtering Tests
 
 You can pass a filter string after the `snforge test` command to filter tests.
-By default, any test with an [absolute module tree path](https://book.cairo-lang.org/ch07-03-paths-for-referring-to-an-item-in-the-module-tree.html#paths-for-referring-to-an-item-in-the-module-tree)
- matching the filter will be run.
+By default, any test with an [absolute module tree path](https://book.cairo-lang.org/ch07-03-paths-for-referring-to-an-item-in-the-module-tree.html#paths-for-referring-to-an-item-in-the-module-tree) matching the filter will be run.
+
+You're tests may differ, note that we changed the default tests generated by `snforge init` for 3 custom tests: `calling`, `calling_another`, and `executing` just for the sake of this example.
```
Comment (kkawula) [inline docs/src/testing/running-tests.md:29]: Is it true? I can't find it

Comment (franciszekjob) [reply docs/src/testing/running-tests.md:29]: In fact I removed this section, I don't think it's worth mentioning at all as it may just confuse user.

File: docs/listings/error_handling/Scarb.toml (hunk 16 above)
Comment (franciszekjob) [reply docs/listings/error_handling/Scarb.toml:12]: We don't have workspaces in listings now (apart from `hello_workspaces` package which is workspace for docs purposes).

File: docs/listings/panicking_test/src/lib.cairo (hunk 15 above)
Comment (franciszekjob) [reply docs/listings/panicking_test/src/lib.cairo:14]: > why this was deleted?

[this section](https://foundry-rs.github.io/starknet-foundry/testing/testing.html#failing-tests) shows the code which should fail, but with this hidden `should_panic` it wouldn't - I think it's misleading.

> shouldn't it be
#[should_panic(expected: 'panic message')]

If we keep it, then yes, but as I said - this example should be failing on purpose.

File: crates/docs/src/utils.rs [hunk 18]
```diff
@@ -0,0 +1,41 @@
+use std::{env, path::PathBuf};
+
+use crate::snippet::Snippet;
+
+#[must_use]
+pub fn get_parent_dir(levels_up: usize) -> PathBuf {
```
Comment (cptartur) [inline crates/docs/src/utils.rs:6]:
```suggestion
pub fn get_nth_ancestor(levels_up: usize) -> PathBuf {
```

Or something different but suggesting it's going up by the number of `levels_up`.

File: crates/docs/src/validation.rs [hunk 19]
```diff
@@ -1,60 +1,54 @@
...
         .captures_iter(&content)
         .filter_map(|caps| {
-            let command_match = caps.get(1)?;
             let match_start = caps.get(0)?.start();
+            let config_str = caps
+                .get(1)
+                .map_or_else(String::new, |m| m.as_str().to_string());
+            let command_match = caps.get(2)?;
+            let output = caps.get(3).map(|m| m.as_str().to_string());
+
```
Comment (cptartur) [inline crates/docs/src/validation.rs:27]: We have everything very nicely abstracted except for this part. You have to know the internal structure of the regex here to know from which capture to get which part

Comment (cptartur) [reply crates/docs/src/validation.rs:27]: Could we at least use named captures for this?

https://docs.rs/regex/latest/regex/struct.Captures.html#method.name

File: crates/forge/tests/e2e/common/runner.rs [hunk 20]
```diff
@@ -33,13 +33,35 @@ pub(crate) fn test_runner(temp_dir: &TempDir) -> SnapboxCommand {
 
 pub(crate) static BASE_FILE_PATTERNS: &[&str] = &["**/*.cairo", "**/*.toml"];
 
+fn is_package_from_docs_listings(package: &str) -> bool {
```
Comment (cptartur) [inline crates/forge/tests/e2e/common/runner.rs:36]: You could [canonicalize](https://doc.rust-lang.org/beta/std/fs/fn.canonicalize.html) both "docs/listings/" directory and given package and check if dir `.starts_with`.

File: crates/forge/tests/e2e/common/runner.rs [hunk 21]
```diff
@@ -33,13 +33,35 @@ pub(crate) fn test_runner(temp_dir: &TempDir) -> SnapboxCommand {
...
     package_name: &str,
     file_patterns: &[&str],
 ) -> TempDir {
     let temp = tempdir_with_tool_versions().unwrap();
-    temp.copy_from(format!("tests/data/{package_name}"), file_patterns)
-        .unwrap();
+
+    let is_from_docs_listings = is_package_from_docs_listings(package_name);
+
+    let package_path = if is_from_docs_listings {
```
Comment (cptartur) [inline crates/forge/tests/e2e/common/runner.rs:51]: Can we maybe inline this?

File: crates/forge/tests/e2e/docs_snippets_validation.rs [hunk 22]
```diff
@@ -1,43 +1,62 @@
...
         };
-        assert_valid_snippet(parse_result.is_ok(), snippet, "snforge", &err_message);
+
+        assert_valid_snippet(parse_result.is_ok(), snippet, &err_message);
+
+        // Remove "snforge" from the args
+        args.remove(0);
+
+        // Remove "test" from the args
+        args.retain(|element| element != &"test");
```
Comment (cptartur) [inline crates/forge/tests/e2e/docs_snippets_validation.rs:44]: What happens if the snippet is `snforge test ...`? Does this just fail or do we handle it somewhere else.

**Review by cptartur — CHANGES_REQUESTED** [review]
Some of the paths used in docs must be invalid because I get errors if I run `mdbook build`.

File: crates/forge/tests/e2e/docs_snippets_validation.rs (hunk 22 above)
Comment (franciszekjob) [reply crates/forge/tests/e2e/docs_snippets_validation.rs]: I updated it to use `runner` instead of `test_runner`.

File: crates/docs/src/snippet.rs [hunk 23]
```diff
@@ -0,0 +1,95 @@
...
+        // <summary>Output:</summary>
+        // ```shell
+        // <output>
+        // ```
+        // </details>
+
+        let escaped_command = regex::escape(self.as_str());
+        let pattern = format!(
+            r"(?ms)^(?:<!--\s*(.*?)\s*-->\n)?```shell\n\$ ({escaped_command} .+?)\n```(?:\s*<details>\n<summary>Output:<\/summary>\n\n```shell\n([\s\S]+?)\n```[\s]*<\/details>)?"
+        );
```
[the remark this answers is not in the document]
Comment (franciszekjob) [reply crates/docs/src/snippet.rs:40]: Alt Text

File: crates/docs/src/snippet.rs [hunk 24]
```diff
@@ -0,0 +1,95 @@
...
+    pub snippet_type: SnippetType,
+    pub config: SnippetConfig,
+}
+
+impl Snippet {
+    pub fn to_command_args(&self) -> Vec<String> {
+        let cleaned_command = self
+            .command
+            .lines()
+            .map(str::trim_end)
```
[the remark this answers is not in the document]
Comment (franciszekjob) [reply crates/docs/src/snippet.rs]: In fact we can get rid of trimming here, as shell-words will take care of that.

File: crates/forge/tests/e2e/common/runner.rs [hunk 25]
```diff
@@ -38,8 +38,26 @@ pub(crate) fn setup_package_with_file_patterns(
...
+    let is_from_docs_listings = fs::read_dir("../../docs/listings")
+        .unwrap()
+        .map(|entry| entry.unwrap().file_name().into_string().unwrap())
+        .any(|entry| entry == package_name);
+
+    let package_path = if is_from_docs_listings {
+        format!("../../docs/listings/{package_name}",)
+    } else {
+        format!("tests/data/{package_name}",)
+    };
```
Comment (cptartur) [inline crates/forge/tests/e2e/common/runner.rs:51]: What I meant when I said "inlining" was this 😅 
```suggestion
    let package_path = if is_from_docs_listings(...) {
        format!("../../docs/listings/{package_name}",)
    } else {
        format!("tests/data/{package_name}",)
    };
```

Not inlining the whole function

**Review by cptartur — APPROVED** [review]
Please address the comment below and one about using `canonicalize`.
Otherwise looks good

**Review by ddoktorski — COMMENTED** [review]
Please re-request review once the failing tests are fixed

File: docs/listings/basic_example/Scarb.toml [hunk 26]
```diff
@@ -6,9 +6,12 @@ edition = "2024_07"
...
 [dependencies]
-snforge_std.workspace = true
-sncast_std.workspace = true
-starknet.workspace = true
+starknet = "2.7.0"
+snforge_std = { path = "../../../snforge_std" }
+sncast_std = { path = "../../../sncast_std" }
+
+[[target.starknet-contract]]
+sncast_std = { path = "../../../sncast_std" }
```
[the remark this answers is not in the document]
Comment (franciszekjob) [reply docs/listings/basic_example/Scarb.toml:14]: Looks like it's not.

File: crates/forge/tests/e2e/common/runner.rs (hunk 25 above)
Comment (franciszekjob) [reply crates/forge/tests/e2e/common/runner.rs:51]: Thne we can't inline it because `is_from_docs_listings` is used [here](https://github.com/foundry-rs/starknet-foundry/blob/dd4ede75a189ef2388b8b037660d5a6ec15d100c/crates/forge/tests/e2e/common/runner.rs#L89) as well.

File: crates/forge/tests/e2e/common/runner.rs (hunk 20 above)
Comment (franciszekjob) [reply crates/forge/tests/e2e/common/runner.rs:36]: I think we can just check if `canonicalize` returns `is_ok`, like this

```rust
fn is_package_from_docs_listings(package: &str) -> bool {
    let package_path = Path::new("../../docs/listings").join(package);
    fs::canonicalize(&package_path).is_ok()
}
```

File: docs/listings/detailed_resources_example/.gitignore
Comment (ddoktorski) [inline docs/listings/detailed_resources_example/.gitignore:1]: This file can be removed. It is enough that we have these entries in the main `.gitignore` file. The same for the other addded `.gitignore` files.

File: crates/forge/tests/e2e/docs_snippets_validation.rs [hunk 27]
```diff
@@ -1,43 +1,59 @@
...
+                .config
+                .package_name
+                .clone()
+                .or_else(|| snippet.capture_package_from_output())
+                .expect("Cannot find package name in command output or snippet config");
+
+            let temp = setup_package(&package_name);
+            let output = runner(&temp).args(args).assert();
+
+            assert_stdout_contains(output, snippet_output);
```
Comment (ddoktorski) [inline crates/forge/tests/e2e/docs_snippets_validation.rs:54]: nit: This could potentially check the exact match

**Review by ddoktorski — APPROVED** [review]

File: crates/forge/tests/e2e/docs_snippets_validation.rs (hunk 27 above)
Comment (franciszekjob) [reply crates/forge/tests/e2e/docs_snippets_validation.rs:54]: In the docs we don't include Scarb's compilation-related output, so exact match wouldn't work here, we just want to check the tests output.
